docs: Add documentation for SRIOV

amshinde opened the pull request:
Most of the documentation has been moved from 2.x with
some cleanup and references to the 3.0 runtime.

Contributions-by: Archana Shinde <archana.m.shinde@intel.com>

Signed-off-by: Archana Shinde <archana.m.shinde@intel.com>

amshinde commented:
@mcastelino Yes, I am working on a separate doc "Passing-devices-to-Clear-Containers.md" that describes steps involved in passing block and vfio device groups. Expect a PR today for that. Thats what you meant right? We dont have support for autodetecting vfio capable devices today.

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,218 @@
+# Setup to use SRIO-V with Clear Containers
+
+Single Root IO Virtualization enables splitting a physical device into
```
- IO or I/O which is more common?

This sentence feels a bit unwieldy to me. How about we split it into something like:

> Single Root I/O Virtualization enables splitting a physical input output device into multiple separate slices, or 
> virtual functions (VFs). This enables direct pass through to virtual machines or containers.

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,218 @@
+
+Single Root IO Virtualization enables splitting a physical device into
+virtual functions (VF) which enables direct pass through to virtual machines or
+containers.  For Clear Containers, we have enabled a CNM plugin and necessary
```
What "necessary changes"? Can we just drop those words?

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,218 @@
+
+## Host setup for SRIO-V
+
+You may need to rebuild your system's kernel in order to disable VFNOIOMMU in
```
- s/your system's kernel/your host system's kernel/
- The kernel config option appears to be `CONFIG_VFIO_NOIOMMU`, not `VFNOIOMMU`.

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,218 @@
+## Host setup for SRIO-V
+
+You may need to rebuild your system's kernel in order to disable VFNOIOMMU in
+the kernel config and potentially add a PCI quirk for your NIC. If not, you can
```
- s/kernel config/kernel configuration/
- The term "NIC" hasn't been defined.

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,218 @@
+
+You may need to rebuild your system's kernel in order to disable VFNOIOMMU in
+the kernel config and potentially add a PCI quirk for your NIC. If not, you can
+move to next section. The below section describes how to assess if
```
- s/to next section/to the next section/
- s/below section/section below/

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,218 @@
+changes are needed, and how to make these changes.
+
+### IOMMU Groups and PCIe Access Control Services
+Taking a look at how the iommu groups are setup on your host system can help
```
s/iommu/IOMMU/ ?

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,218 @@
+### IOMMU Groups and PCIe Access Control Services
+Taking a look at how the iommu groups are setup on your host system can help
+provide information on whether or not your NIC is setup appropriately with
+respect to PCIe Access Control Services. More specifically, if the PCI bridge is
```
s/Access Control Services/Access Control Services (ACS)/.

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,218 @@
+Taking a look at how the iommu groups are setup on your host system can help
+provide information on whether or not your NIC is setup appropriately with
+respect to PCIe Access Control Services. More specifically, if the PCI bridge is
+within the same iommu-group as your NIC, then this is an indication that either
```
s/iommu-group/IOMMU group/ ?

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,218 @@
+update your host's kernel config to disable VFNOIOMMU.
+
+```
+find /sys/kernel/iommu_groups/ -type l
````
Needs a leading `$ ` to denote what follows is a command the user needs to type.

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,218 @@
+appropriately by default.
+
+For example, when you run the following command, if all is setup properly, you
+should have the PCI for each ACS enabled NIC port in its own iommu_group. If
```
iommu-group has now become iommu_group. Again, should this be "IOMMU group"?

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,218 @@
+find /sys/kernel/iommu_groups/ -type l
+```
+
+For more details, see http://vfio.blogspot.com/2014/08/iommu-groups-inside-and-out.html
````
This doesn't appear to be an official standards or reference site so I'm not sure it should be included.

@iphutch - can you comment please?

amshinde replied:
@jodh-intel This is the blog of the kernel maintainer for VFIO, with all the information related to VFIO.

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,218 @@
+### Updating the Host Kernel
+
+These directions are provided as an example, and are based on Ubuntu 16.04. The
+user will need to do equivalent for their distribution.
```
/will need to do equivalent/will need to do something equivalent/ ?

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,218 @@
+1.  Grab kernel sources
+
+```
+sudo apt-get install linux-source-4.10.0
````
Again, all these commands should have a `$ ` prefix.

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,218 @@
+2. Check the config and update if necessary
+
+```
+sudo cp /boot/config-4.8.0-36-generic .config
````
`$ `.

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,218 @@
+
+```
+sudo cp /boot/config-4.8.0-36-generic .config
+sudo make olddefconfig #and verify resulting .config does not have NOIOMMU set; ie: # CONFIG_VFIO_NOIOMMU is not set
````
Can you move this comment to a separate line above the command to run. Also, please can you quote the config string so it shows as `"# CONFIG_VFIO_NOIOMMU is not set"`.

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,218 @@
+
+3.  If necessary, add PCI Quirk for SRIOV NIC
+
+Now, depending on how your NIC describes its ACS capabilities, you may need to add
```
Remove "Now, ".

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,218 @@
+```
+modify drivers/pci/quirks.c:
+line 4118:
+static const u16 pci_quirk_intel_pch_acs_ids[] = {
````
We're suggesting users change the contents of their host kernel. This needs two things imho:

- A warning explaining the risks of doing this (YMMV, "you get to keep the pieces...", etc).
- A suggestion that users also add a unique identifier to the `EXTRAVERSION` variable so that when the kernel is installed and running, running `uname -r` will show that this is a custom kernel.
- A warning explaining that if your system is configured for automatic distribution updates, you may lose this extra functionality the next time you run `dnf update`, `apt-get update`, etc.

egernst replied:
Archana, on your system did you need to make this change?  I think if this is a formal documentation / how-to, perhaps we shouldn't suggest this.  This'll make it look like the chipset supports true IOMMU, which may not be the case if you hack this driver...  perhaps we should just skip this, as it shouldn't be a normal occurrence for users...

egernst replied:
@amshinde ^

amshinde replied:
@egernst No, I didnt need to do this, I was testing mainly on the test machine we have set up, where the 2 NIC ports were in separate IOMMU groups. I assumed you had to do this for your setup.
This is not formal documentation, but the kernel maintainer suggests  the approach to apply the patch here:
https://lkml.org/lkml/2013/5/30/513
 to quirk specific devices which enforce isolation but not provide an ACS capability. Note the vendor needs to confirm that the device supports isolation. I think I am going to take this out completely and instead add a link to to the pci-override patch , stating that this should be done only in case the device supports isolation.

egernst replied:
@amshinde - ok.  I think this should be rare, and only needed if they have an out of date kernel.  The quirk should be upstream already if their device doesn't advertise ACS correctly but does indeed support hardware isolation.  Worse case is they'd be marking it as supporting IOMMU when it indeed does not.  The link sounds best way to describe this rare case.

amshinde replied:
@egernst Changes pushed for this.

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,218 @@
+For SRIOV, a physical device can create up to `sriov_totalvfs` virtual 
+functions (VFs).  Once created, you cannot grow or shrink the number of VFs 
+without first setting it back to zero.  Based on this, it is expected that you
+should set the number of VFs from a physical device  just once.
```
s/device  just/device just/

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,218 @@
+The following will be needed per boot in order to facilitate setting up a
+physical device's virtual functions.
+
+For SRIOV, a physical device can create up to `sriov_totalvfs` virtual 
```
`sriov_totalvfs` hasn't been introduced yet - it's explained below so this needs to be mentioned here.

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,218 @@
+```
+4.  Create VFs:
+
+Create virtual functions by editing sriov_numvfs. In our example, let's 
````
`sriov_numvfs` is a file so should be rendered in a fixed width font.

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,218 @@
+For example,
+
+```
+ip link set <pf> vf <vfidx> mac <my made up mac address>
````
Might be better to change "my made up mac address" to simply "fake MAC address".

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,218 @@
+
+1.  Build and start SRIOV plugin
+
+This assumes you already have GOPATH set in your environment.
```
`GOPATH`.

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,218 @@
+
+This assumes you already have GOPATH set in your environment.
+```
+sriov@sriov-2:~$ sudo mkdir /etc/docker/plugins
````
`mkdir -p`.

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,218 @@
+```
+sriov@sriov-2:~$ sudo mkdir /etc/docker/plugins
+sriov@sriov-2:~$ sudo cp go/src/github.com/clearcontainers/sriov/sriov.json /etc/docker/plugins/
+cd /go/src/github.com/clearcontainers/sriov
````
Shouldn't both these lines reference `$GOPATH/src/github.com/...`?

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,218 @@
+look at 128 virtual devices.
+
+```
+root@sriov-1:/home/sriov# echo 1 > /sys/bus/pci/devices/0000\:01\:00.0/sriov_numvfs
````
Can you change the prompt from `root@sriov-1:/home/sriov# ` to `# ` for clarity (here and elsewhere)?

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,218 @@
+
+Machine #1:
+```
+sriov@sriov-2:~$ sudo docker run --runtime=cc-runtime --net=vfnet --ip=192.168.0.10 -it mcastelino/iperf iperf3 -s
````
We need to explain why the `mcastelino/iperf` docker image is being used here. Is it required? If so, why?

amshinde replied:
@mcastelino what do you think about this?
You mentioned today about having a Docker image with a Dockerfile with all the necessary tools.

amshinde commented:
@jodh-intel thanks for taking a look. I have pushed changes for your comments.

amshinde commented:
@iphutch Can you take a look at this?

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,239 @@
+$ sudo make olddefconfig 
+```
+
+You may choose to modify the kernel Makefile to add a unique indentifier 
````
`Makefile`.

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,239 @@
+kernel being run has been customized.
+If your system is configured for automatic distribution updates, make sure
+you are running your customized kernel and not the new kernel
+installed with apt-get update or dnf update.
```
`apt-get`, `dnf`.

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,239 @@
+
+Depending on how your NIC describes its ACS capabilities, you may need to add
+a quirk in order to indicate that the given NIC does properly support ACS.  An
+example is given below, but your mileage will vary (at the very least, check your
```
I'd expand the description by saying something like:

> The example below shows the `diff(1)` output for adding a new quirk for PCI ID 0x0c01.
> ```
> $ diff drivers/pci/quirks.c
> <Add diff output here>
> ```

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,239 @@
+4.  Build and install the kernel:
+
+```
+sudo make -j #and go get coffee
````
These commands and all below still need to have a `$ ` prompt.

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,239 @@
+
+6.  Reboot and verify
+Host system should be ready now -- reboot and verify that expected cmdline and
+kernel version is booted (look at /proc/cmdline and /proc/version)
```
`/proc/cmdline`, `/proc/version`.

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,237 @@
+3.  Build and install the kernel:
+
+```
+$ sudo make -j #and go get coffee
````
- I think you should drop the `sudo` for this line and the one below - it should only be used for the installation steps.
- Clearly, `-j` helps speed up the build, but does tend to completely DoS a machine. I'd be inclined to make this `-j3` for parity with the line below.

/me hopes the comment isn't "beveragist" too ;)

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,237 @@
+$ sudo make install
+```
+
+4.  Edit grub to enable intel-iommu:
````
Should that be `intel_iommu`?

iphutch replied:
+1

jodh-intel commented:
Hi @iphutch - could you take a look at this?

iphutch commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,237 @@
+
+## Host setup for SR-IOV
+
+You may need to rebuild your host system's kernel in order to disable 
```
Am I right in saying we have the following different ideas/purposes for this section:
1) Introducing that `CONFIG_VFIO_NOIOMMU` needs to be disabled
2) Informing readers that they may need to rebuild the kernel disable `CONFIG_VFIO_NOIOMMU`
3) *Instructing readers on how to disable `CONFIG_VFIO_NOIOMMU` if no rebuild is necessary*
4) Informing readers that adding a PCI may be necessary and introducing the section where the reader can find out if that is necessary.

If so, we need to:
* Add to the beginning of this section: `CONFIG_VFIO_NOIOMMU` must be disabled to setup your host   system for SR-IOV. The method for disabling `CONFIG_VFIO_NOIOMMU` differs depending on the host system's configuration. 
* Separate the sentence saying the host's system kernel needs to be rebuild and a PCI needs to be added. 
* Add instructions on how to disable `CONFIG_VFIO_NOIOMMU` when you do not need to rebuild the kernel.

amshinde replied:
I have changed to wording here.

iphutch commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,237 @@
+changes are needed, and how to make these changes.
+
+### IOMMU Groups and PCIe Access Control Services
+Taking a look at how the Input-Output Memory Management Unit (IOMMU) groups are
```
Add a blank line after the heading

iphutch commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,237 @@
+NIC is setup appropriately with respect to PCIe Access Control Services (ACS).
+More specifically, if the PCI bridge is within the same IOMMU group as your 
+NIC, then this is an indication that either your device doesn't support ACS or 
+that it doesn't share this capability appropriately by default.
```
This section is written in passive voice. Let's change it to be more direct:

Run the following command to see how the Input-Output Memory Management Unit (IOMMU) groups are setup on your host system. This information details whether or not your NIC is setup appropriately with respect to PCIe Access Control Services (ACS). If the PCI bridge is within the same IOMMU group as your NIC, then this is an indication that either your device doesn't support ACS or that it doesn't share this capability appropriately by default.

```
$ find /sys/kernel/iommu_groups/ -type l
```

If all if all is setup properly, you will have the PCI for each ACS enabled NIC port in its own IOMMU group. If you don't see any output when running the following, then you likely need to update your host's kernel configuration to disable `CONFIG_VFIO_NOIOMMU`.

For more details, see http://vfio.blogspot.com/2014/08/iommu-groups-inside-and-out.html

iphutch commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,237 @@
+$ sudo apt-get install libssl-dev
+```
+
+2. Check the config and update if necessary
````
`config` file

iphutch commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,237 @@
+$ sudo make olddefconfig 
+```
+
+You may choose to modify the kernel `Makefile` to add a unique indentifier 
````
Can we add a reason they would they choose that? If it is recommended, lets just say: To modify..

amshinde replied:
The reason is mentioned here , so that when the kernel is running, the command `uname -r` indicates a customised kernel is being run.

iphutch commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,237 @@
+
+You may choose to modify the kernel `Makefile` to add a unique indentifier 
+to the `EXTRAVERSION` variable proir to running make, so that when the
+kernel is installed and running, the command `uname -r` will indicate that the
```
replace/will indicate/indicates

iphutch commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,237 @@
+```
+
+5.  Reboot and verify
+Host system should be ready now -- reboot and verify that expected cmdline and
````
add blank line and indent all that belong to numbered steps.
substitute/that expected cmdline/that the expected command line

iphutch commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,237 @@
+These directions are provided as an example, and are based on Ubuntu 16.04. The
+user will need to do something equivalent for their distribution.
+
+1.  Grab kernel sources
```
We need to indent all content following a numbered step in this section.

iphutch commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,237 @@
+sudo reboot
+```
+
+6. Add `vfio-pci` module
````
substitute: Add the `vfio-pci`...
Add a blank line after the numbered step and intent

iphutch commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,237 @@
+
+7.  If necessary, add PCI quirk for SR-IOV NIC
+
+Verify your NIC appears in its own IOMMU group and no ther devices 
```
Indent

iphutch commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,237 @@
+
+## Setting up SR-IOV Device
+
+All of the prior sections are needed once to prepare the SR-IOV host systems.
```
remove/once

iphutch commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,237 @@
+## Setting up SR-IOV Device
+
+All of the prior sections are needed once to prepare the SR-IOV host systems.
+The following will be needed per boot in order to facilitate setting up a
```
I don't understand "per boot" here

amshinde replied:
It means every time system is booted. Changed this to per system boot.

iphutch commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,237 @@
+The following will be needed per boot in order to facilitate setting up a
+physical device's virtual functions.
+
+For SR-IOV, a physical device can create up to `sriov_totalvfs` virtual 
```
Let's put the explanation of `sriov_totalvfs` before this sentence as currently: create up to `sriov_totalvfs` doesn't make sense

iphutch commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,237 @@
+physical device's virtual functions.
+
+For SR-IOV, a physical device can create up to `sriov_totalvfs` virtual 
+functions (VFs). The `sriov_totalvfs` file under sysfs for a pci device
```
under the `sysfs` directory

iphutch commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,237 @@
+For SR-IOV, a physical device can create up to `sriov_totalvfs` virtual 
+functions (VFs). The `sriov_totalvfs` file under sysfs for a pci device
+specifies the total number of available virtual functions. Once VF's are created, 
+you cannot grow or shrink the number of VFs without first setting it back 
```
substitute/it/the number

iphutch commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,237 @@
+to zero. Based on this, it is expected that you should set the number of VFs 
+for a physical device just once.
+
+1. Add `vfio-pci` device driver
```
indent all content that falls under a numbered step

iphutch commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,237 @@
+01:00.1 Ethernet controller: Intel Corporation Ethernet Controller 10-Gigabit X540-AT2 (rev 01)
+```
+In our case, both 01:00.0 and 01:00.1 are the two ports on our x540-AT2 card 
+that we'll use.  You can use lshw to get further details on the controller 
````
is the abbreviation lshw written out prior to this?

iphutch commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,237 @@
+and verify it indeed supports SR-IOV.
+
+3.  Check how many VFs we can create:
+```
````
add blank line after heading

iphutch commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,237 @@
+```
+4.  Create VFs:
+
+Create virtual functions by editing `sriov_numvfs`. In our example, let's 
````
substitute/let's just create/we create

iphutch commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,237 @@
+6.  Assign MAC address to each VF
+
+Depending on the NIC being used, you may need to explicitly set the MAC
+address for the Virtual Function device in order to guarantee that the 
```
use lower case for virtual function or utilize abbreviation

iphutch commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,237 @@
+
+1.  Build and start SR-IOV plugin
+
+This assumes you already have `GOPATH` set in your environment.
```
indent

iphutch commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,237 @@
+```
+2.  Create docker network
+
+To create the SR-IOV based docker network, a subnet, vlanid and physical interface are required.
````
vlanid,

iphutch commented:
@klynnrif could you please take a look?

klynnrif commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,265 @@
+# Setup to use SR-IOV with Clear Containers
```
Capitalize "Use"

klynnrif commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,265 @@
+   that we'll use.  You can use `lshw` command to get further details on the 
+   controller and verify it indeed supports SR-IOV.
+
+3. Check how many VFs we can create:
```
Check how many VFs you can create:

klynnrif commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,265 @@
+   $ cat /sys/bus/pci/devices/0000\:01\:00.1/sriov_totalvfs
+   63
+   ```
+
````
Add the following sentence after the two cat commands: 

The previous commands show how many VFs you can create. The `sriov_totalvfs` file under `sysfs` for a PCI device specifies the total number of VFs that you can create.

klynnrif commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,265 @@
+   63
+   ```
+
+4. Create VFs:
````
Lines 192-201 rewrite: 

Create the VFs: 

NOTE: insert command immediately following numbered step

The previous commands create virtual functions by editing `sriov_numvfs`. This example creates one VF per physical device. Note, creating one VF eliminates the usefulness of SR-IOV, and is done for simplicity in this example.

klynnrif commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,265 @@
+   # echo 1 > /sys/bus/pci/devices/0000\:01\:00.1/sriov_numvfs
+   ```
+
+5. Verify that these were added to the host:
````
Verify that the VFs were added to the host:

klynnrif commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,265 @@
+   02:10.1 Ethernet controller: Intel Corporation X540 Ethernet Controller Virtual Function (rev 01)
+   ```
+
+6. Assign MAC address to each VF
````
Lines 212-220 rewrite: 

Assign a MAC address to each VF: 

NOTE: insert command immediately following numbered step

Depending on the NIC being used, you  might need to explicitly set the MAC address for the VF device. Setting the MAC address guarantees that the address is consistent on the host and when passed to the guest.

klynnrif commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,265 @@
+   $ ip link set <pf> vf <vfidx> mac <fake MAC address>
+   ```
+
+## Example: Launch a Clear Container using SR-IOV
````
Example: Launching a Clear Container Using SR-IOV

The following example launches a Clear Container using SR-IOV:

klynnrif commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,265 @@
+
+## Example: Launch a Clear Container using SR-IOV
+
+1. Build and start SR-IOV plugin
```
Lines 225-233 rewrite: 

Build and start the SR-IOV plugin: 

NOTE: insert command immediately following numbered step

The previous commands assumes you have already set `GOPATH` in your environment.

klynnrif commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,265 @@
+   $ sudo ./sriov &
+   ```
+
+2. Create docker network
````
Lines 236-244 rewrite: 

Create the docker network: 

NOTE: insert command immediately following numbered step

The previous commands create the SR-IOV docker network, a subnet, vlanid, and a physical interface, all which are required.

klynnrif commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,265 @@
+   ee2e5a594f9e4d3796eda972f3b46e52342aea04cbae8e5eac9b2dd6ff37b067
+   ```
+
+3. Start containers and test connectivity
````
Lines 247-251 rewrite: 

Start the containers and test their connectivity: 

NOTE: insert sudo docker run command immediately following numbered step

The previous command starts a container, which uses SR-IOV.

klynnrif commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,265 @@
+   $ sudo docker run --runtime=cc-runtime --net=vfnet --ip=192.168.0.10 -it alpine
+   ```
+
+   If you have two machines with SR-IOV enabled NICs connected back to back and 
````
Lines 254-256 rewrite: 

The following two commands test the connectivity when two machines with SR-IOV-enabled NICs are connected back-to-back and each machine has a network with a matching Virtual LAN ID (vlan-id):

amshinde replied:
This makes it less clear I think.

klynnrif replied:
Thanks @amshinde, how's this?    If two machines with SR-IOV enabled NICs are connected back-to-back and each has a network with matching vlan-id created, use the following two commands to test the connectivity:

grahamwhaley commented:
@klynnrif Ah, I see you used the 'Review Changes' method to group comments here - many thanks! (and for the reviews in general ;-) )

rcaballeromx commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,265 @@
+
+## Host setup for SR-IOV
+
+You need to have a system with Intel VT-d support. Your kernel needs to have 
```
I agree. The introductory sentence should read: "To setup your host for SR-IOV, the ensure the following requirements are met:"
The first bullet should read: "The host system must support Intel VT-d."
The third bullet should read: "Ensure the host kernel's configuration for SR-IOV, `CONFIG_VFIO_NOIOMMU` is disabled. You will need to rebuild your host system's kernel to disable it.

rcaballeromx commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,265 @@
+
+Run the following command to see how the IOMMU groups are setup on your 
+host system. This information details whether or not your NIC is setup 
+appropriately with respect to PCIe Access Control Services (ACS). If the PCI 
```
I agree but the conditional is still needed: "When the PCI bridge is within the same IOMMU group as your NIC, then your device does not support ACS or does not share this capability by default."

amshinde replied:
We are indicating a possible situation here, hence the use of "if"

rcaballeromx commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,265 @@
+$ find /sys/kernel/iommu_groups/ -type l
+```
+
+If all if all is setup properly, you will have the PCI for each ACS enabled 
````
I again believe the conditional is needed: "When the IOMMU groups are set up properly, the PCI for each ACS-enabled NIC port is within its own IOMMU group."

rcaballeromx commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,265 @@
+```
+
+If all if all is setup properly, you will have the PCI for each ACS enabled 
+NIC port in its own IOMMU group. If you don't see any output when running the above
````
I agree with the new paragraph but the sentence should also be simplified: 
If the previous command produces no output, then you should to update your host's kernel configuration.

rcaballeromx commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,265 @@
+NIC port in its own IOMMU group. If you don't see any output when running the above
+command, then you likely need to update your host's kernel configuration.
+
+For more details, see http://vfio.blogspot.com/2014/08/iommu-groups-inside-and-out.html
```
Adding a colon after see is better. Some browsers have problems with periods at the end of links.

rcaballeromx commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,265 @@
+
+For more details, see http://vfio.blogspot.com/2014/08/iommu-groups-inside-and-out.html
+
+### Updating the Host Kernel
```
Use infinitives for headings: Update the host kernel
Make sure all headings use sentence case: Only the first word and proper nouns (names) are capitalized. 
This applies to all headings.

rcaballeromx commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,265 @@
+configuration. If either of the above checks fail, you will need to build a 
+new kernel following steps 1 through 3 listed below.
+
+The directions below are provided as an example, and are based on Ubuntu 16.04. You
```
Keep content short and to the point. Avoid using complex words for simple actions:  The following steps update the SR-IOV host system. The steps are based on the Ubuntu 16.04 distribution. If you use a different distribution, make the appropriate adjustments to the commands.

rcaballeromx commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,265 @@
+The directions below are provided as an example, and are based on Ubuntu 16.04. You
+will need to do something equivalent for their distribution.
+
+1. Grab kernel sources
```
Simplify: To get the kernel sources, enter the following commands:

rcaballeromx commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,265 @@
+   $ sudo apt-get install libssl-dev
+   ```
+
+2. Check the `config` file and update if necessary
````
Split this step in two:
To verify the `config` file, enter the following command:
To update the `config` file, enter the following command:
You will also need to split the two commands and place them where appropriate.

klynnrif replied:
Good to split - question to confirm: It seems that the first three lines here are to create a .config based on a copy command, and then to hand-examine the .config file to make sure what's in the commented lines is true. Your suggestion to "enter the following command" is much simpler than that - maybe a different introductory sentence for the first three lines?

rcaballeromx commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,265 @@
+   $ sudo make olddefconfig 
+   ```
+
+   You may choose to modify the kernel `Makefile` to add a unique indentifier 
````
Simplify: To add a unique identifier to the `EXTRAVERSION` variable prior to running make, modify the kernel's `Makefile`. When you include the `EXTRAVERSION` variable, the `uname -r` command shows that a customized kernel is installed and running.

rcaballeromx commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,265 @@
+   sudo update-grub
+   ```
+
+5. Reboot and verify
````
To reboot the kernel, enter the following command:

rcaballeromx commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,265 @@
+
+5. Reboot and verify
+
+   Host system should be ready now -- reboot and verify that the expected 
```
The `sudo reboot` command does not verify anything. Add an extra step for the verification:
6. To verify the versions of the terminal and the kernel, open the `/proc/cmdline` file and the `/proc/version` file respectively.

rcaballeromx commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,265 @@
+   sudo reboot
+   ```
+
+6. Verify Intel VT-d has been initialized. Check for the following line in 
````
To check if Intel VT-d initialized correctly, look for the following line in the `dmesg` output:

rcaballeromx commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,265 @@
+   DMAR: Intel(R) Virtualization Technology for Directed I/O
+   ```
+
+   On older kernels, the prefix is different.
````
What prefix? Keep sentence structures as simple as possible: Older kernels use a different prefix.

klynnrif replied:
In this case the prefix is "PCI-DMA", which is shown after the sentence. What about, "Older kernels use a different prefix (e.g. PCI-DMA)"?

rcaballeromx commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,265 @@
+   PCI-DMA: Intel(R) Virtualization Technology for Directed I/O
+   ```
+
+7. Add `vfio-pci` module
````
Consistency: To add the `vfio-pci` module, enter the following command:

amshinde commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,265 @@
+
+Single Root I/O Virtualization (SR-IOV) enables splitting a physical device into
+virtual functions (VFs). This enables direct passthrough to virtual machines or
+containers.  For Clear Containers, we have enabled a Container Network Model (CNM)
```
This change changes the entire meaning and makes it inaccurate.

klynnrif replied:
Okay, thanks!

rcaballeromx replied:
@amshinde Please make sure the meaning remains but you still use active voice and ensure the subject is always clear. The 'this' on line four is ambiguous and a reader cannot clearly identify the reference.

amshinde commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,265 @@
+# Setup to use SR-IOV with Clear Containers
```
I'd rather keep this as an action item that the user needs to perform.

klynnrif replied:
With work I've done on the Yocto Project, we use gerund form for headings so that came through here :) Thank you for the heads up that you not use gerund forms of verbs here - I'll make note of that for the future. Given the original heading "Setup to use SR-IOV..." I would suggest changing to "Set up to use SR-IOV..." because it is a set of tasks. I defer to @rcaballeromx

amshinde commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,265 @@
+container's network namespace to facilitate making use of SR-IOV for network 
+based devices.
+
+## Install the SR-IOV Docker\* plugin
```
Same as above

amshinde commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,265 @@
+# Setup to use SR-IOV with Clear Containers
+
+Single Root I/O Virtualization (SR-IOV) enables splitting a physical device into
+virtual functions (VFs). This enables direct passthrough to virtual machines or
```
What do you gain here. I think the previous keeps the sentence short and clearer.

jodh-intel replied:
I think we've already discussed this sentence: with a comma, it feels a bit "unwieldy" to me as it's (a) quite long, and (b) introduces two terms that take a bit of thinking about.

I do feel that keeping it as currently written is clearer as the period gives you time to consider before moving on to the next one. But it's just my opinion so will defer to the experts.

klynnrif replied:
Agreed that the comma makes the sentence quite long. Optional change to get rid the demonstrative pronoun, "This":  ...virtual functions (VSs). Splitting a physical device enables direct passthrough to virtual machines or containers. Although, this may change the entire meaning.

rcaballeromx replied:
Getting rid of demonstrative pronouns is always a win in my book. +1

amshinde commented:
@rcaballeromx @iphutch @klynnrif I have addressed some of your feedback. I have not addressed some of the comments that change the technical accuracy. If there are some changes that you think need to be done lets open a separate PR for it, as the comments on this PR are causing the page to not load for me on multiple occasions and I would like to get some documentation merged in for users interested in SRIOV.

jodh-intel commented:
@amshinde - again I see changes that need to be applied that don't change the technical content.

Also, I think pullapprove is confused because the author is Eric but the sign-off is from you. Maybe if @egernst is happy you could change the commit ownership to you and add a Contributions-by for him to appease the pullapprove Gods?

egernst commented:
@jodh-intel @amshinde -- fine by me

amshinde commented:
@klynnrif Can you take a look at this PR and approve if this looks good to you.

amshinde commented:
@klynnrif Can you take a look?

amshinde commented:
@klynnrif @rcaballeromx Can you take a look please?

klynnrif commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,281 @@
+   and no other devices appear in the same group. In the rare case where your 
+   PCI NIC does not appear in its own group, the NIC might not support ACS or 
+   you may have chosen to build and run an old kernel. Depending on your NIC 
+   and if it indeed enforces isolation, you might resolve this by adding a 
```
and if it enforces isolation...

klynnrif commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,281 @@
+   you may have chosen to build and run an old kernel. Depending on your NIC 
+   and if it indeed enforces isolation, you might resolve this by adding a 
+   `pcie_acs_override=` option to your kernel command line and reboot. 
+   You can find detailed information about this option at [PCIE-ACS-override-option](https://lkml.org/lkml/2013/5/30/513).
```
See [PCIE-ACS-override-option](https://lkml.org/lkml/2013/5/30/513) for detailed information about this option.

klynnrif commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,281 @@
+
+See the following steps one through three if you need to rebuild the kernel.
+
+The following steps update the SR-IOV host system. The steps are based on 
```
Lines 69-71 suggested rewrite: The following steps, which are based on the Ubuntu 16.04 distribution, update the SR-IOV host system. If you use a different distribution, make appropriate adjustments to the commands:

klynnrif commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,281 @@
+   ```
+   The previous command verifies that your NIC appears in its own IOMMU group
+   and no other devices appear in the same group. In the rare case where your 
+   PCI NIC does not appear in its own group, the NIC might not support ACS or 
````
Lines 152-153 suggested rewrite: PCI NIC does not appear in its own group, it is likely that the NIC does not support ACS or you built and ran an old kernel.

klynnrif commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,281 @@
+virtual functions (VF) you can create, and creating those virtual functions.
+Once you create VFs you cannot increase or decrease the number of VFs without 
+first setting the number back to zero. Based on this, it is expected that you
+should set the number of VFs for a physical device just once.
```
Suggest removing the word "should": it is expected that you set the number...

klynnrif commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,281 @@
+
+1. Add `vfio-pci` device driver
+
+   `vfio-pci` is a driver which is used to reserve a VF PCI device.  Add it:
```
... is a driver used to reserve...

klynnrif commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,281 @@
+   01:00.1 Ethernet controller: Intel Corporation Ethernet Controller 10-Gigabit X540-AT2 (rev 01)
+   ```
+   In our case, both 01:00.0 and 01:00.1 are the two ports on our x540-AT2 card 
+   that we'll use.  You can use `lshw` command to get further details on the 
````
that we will use.

klynnrif commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,281 @@
+   ```
+   In our case, both 01:00.0 and 01:00.1 are the two ports on our x540-AT2 card 
+   that we'll use.  You can use `lshw` command to get further details on the 
+   controller and verify it indeed supports SR-IOV.
````
controller and verify it supports SR-IOV.

klynnrif commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,281 @@
+   each has a network with matching vlan-id created, you can test connectivity
+   between them as follows:
+
+   If two machines with SR-IOV enabled NICs are connected back-to-back and each 
```
+1 assuming you will delete the prior paragraph.

amshinde commented:
@klynnrif I have addressed your comments, please take a look.

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
```diff
@@ -0,0 +1,278 @@
+   has a network with matching vlan-id created, use the following two commands 
+   to test the connectivity:
+   
+   Machine #1:
```
Nit: I think "Machine 1" and "Machine 2" would be better.

amshinde replied:
fixed

jodh-intel commented on docs/Using-SRIOV-and-Clear-Containers.md:
````diff
@@ -0,0 +1,278 @@
+   
+   Machine #1:
+   ```
+   sriov@sriov-2:~$ sudo docker run --runtime=cc-runtime --net=vfnet --ip=192.168.0.10 -it mcastelino/iperf iperf3 -s
````
Nits:

- If this is "Machine 1", you'd expect the prompt to be `sriov@sriov-1:~$`, not `sriov@sriov-2:~$`.
- These prompts are a bit complex - wouldn't, `sriov-1:$ ` and `sriov-2:$ ` suffice?

amshinde replied:
fixed

jodh-intel commented:
Thanks @amshinde.

@rcaballeromx - are you happy with this now?

jodh-intel commented:
Hi @rcaballeromx - Happy New Year! Are you happy for this to be merged?

jodh-intel commented:
Ping @rcaballeromx.
